lib/Runtime/Language/JavascriptOperators.cpp

```diff
@@ -10449,18 +10459,30 @@ namespace Js
     {
         Var func = JavascriptOperators::GetProperty(iterator, PropertyIds::next, scriptContext);
 
+        ThreadContext *threadContext = scriptContext->GetThreadContext();
         if (!JavascriptConversion::IsCallable(func))
         {
+            if (!threadContext->RecordImplicitException())
+            {
+                return scriptContext->GetLibrary()->GetUndefined();
+            }
             JavascriptError::ThrowTypeError(scriptContext, JSERR_NeedFunction);
         }
 
         RecyclableObject* callable = RecyclableObject::FromVar(func);
         Js::Var args[] = { iterator, value };
         Js::CallInfo callInfo(Js::CallFlags_Value, _countof(args) + (value == nullptr ? -1 : 0));
-        Var result = JavascriptFunction::CallFunction<true>(callable, callable->GetEntryPoint(), Js::Arguments(callInfo, args));
+        Var result = threadContext->ExecuteImplicitCall(callable, ImplicitCall_Accessor, [&]() -> Var
+            {
```

Comment:
No, this is intended. The C++ code passed to ExecuteImplicitCall should be the whole action you want to take if implicit calls are permitted, or avoid doing if they’re not. In some cases, there’s some checking/setup that only makes sense to do if you’re going to go ahead and call the user method, but here there isn’t much. (Though I suppose the one CallInfo creation could be moved into the ExecuteImplicitCall block.)

--Paul

From: Akrosh Gandhi [mailto:notifications@github.com]
Sent: Thursday, June 16, 2016 11:27 AM
To: Microsoft/ChakraCore <ChakraCore@noreply.github.com>
Cc: Paul Leathers <pleath@microsoft.com>; Author <author@noreply.github.com>
Subject: Re: [Microsoft/ChakraCore] Calls to overridable iteration methods should go through ExecuteImplicitCall so that side-effects are tracked if we wind up executing user code. (#1147)


In lib/Runtime/Language/JavascriptOperators.cpp<https://github.com/Microsoft/ChakraCore/pull/1147#discussion_r67398672>:

>              JavascriptError::ThrowTypeError(scriptContext, JSERR_NeedFunction);

>          }

>

>          RecyclableObject* callable = RecyclableObject::FromVar(func);

>          Js::Var args[] = { iterator, value };

>          Js::CallInfo callInfo(Js::CallFlags_Value, _countof(args) + (value == nullptr ? -1 : 0));

> -        Var result = JavascriptFunction::CallFunction<true>(callable, callable->GetEntryPoint(), Js::Arguments(callInfo, args));

> +        Var result = threadContext->ExecuteImplicitCall(callable, ImplicitCall_Accessor, [&]() -> Var

> +            {

Also in this file - I have seen few other places we have ExecuteImplicitCall, but the whole function is wrapped under ExecuteImplicitCall, as opposed to what you have done (just the callfunction). Was that oversight?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/Microsoft/ChakraCore/pull/1147/files/d99e4a8d54c2203aa602e35744e29d38728acda4#r67398672>, or mute the thread<https://github.com/notifications/unsubscribe/APF8RJYHenVHWSTeEdra5cLAKfRlWCxDks5qMZVggaJpZM4I28iQ>.